src/shred_network/shred_receiver.zig

```diff
@@ -186,13 +201,338 @@ pub const ShredReceiver = struct {
+        }
+    }
+
+    fn handleDuplicateSlot(
+        self: *ShredReceiver,
+        sender: anytype,
+        slot: Slot,
+        shred_payload: []const u8,
+        duplicate_payload: []const u8,
+    ) !void {
+        // Store in ledger
+        self.params.result_writer.storeDuplicateSlot(
+            slot,
+            shred_payload,
+            duplicate_payload,
+        ) catch |err| {
+            self.logger.err().logf(
+                "failed to store duplicate slot {}: {}",
+                .{ slot, err },
+            );
+        };
+
+        // Send to consensus
+        sender.send(slot) catch |err| {
+            self.logger.err().logf(
+                "failed to send duplicate slot {} to consensus: {}",
+                .{ slot, err },
+            );
+        };
+
+        // Broadcast duplicate shred proof via gossip
+        if (self.params.gossip_service) |gossip| {
+            self.pushDuplicateShredToGossip(
+                gossip,
+                slot,
+                shred_payload,
+                duplicate_payload,
+            ) catch |err| {
+                self.logger.err().logf(
+                    "failed to push duplicate shred to gossip for slot {}: {}",
+                    .{ slot, err },
+                );
+            };
+        }
+    }
+
+    fn pushDuplicateShredToGossip(
+        self: *ShredReceiver,
+        gossip: *sig.gossip.GossipService,
+        slot: Slot,
+        shred_payload: []const u8,
+        other_payload: []const u8,
+    ) !void {
+        const my_pubkey = sig.core.Pubkey.fromPublicKey(&self.params.keypair.public_key);
+
+        // Early return if we already have a duplicate for this slot.
+        if (hasDuplicateForSlot(gossip, my_pubkey, slot)) {
+            return;
+        }
+
+        // Compute ring offset where new entries should be placed/overwritten.
+        const ring_offset = computeRingOffset(gossip, my_pubkey);
+
+        // Serialize duplicate slot proof.
+        const allocator = gossip.allocator;
+        const proof_bytes = try serializeDuplicateProof(allocator, shred_payload, other_payload);
+        defer allocator.free(proof_bytes);
+
+        // Build chunks that will be converted to CRDS.
+        const chunks = try self.buildDuplicateShredChunks(
+            gossip,
+            slot,
+            shred_payload,
+            proof_bytes,
+            @as(usize, DUPLICATE_SHRED_MAX_PAYLOAD_SIZE),
+        );
+        defer {
+            for (chunks) |dup| gossip.allocator.free(dup.chunk);
+            gossip.allocator.free(chunks);
+        }
+
+        try enqueueDuplicateShredCrdsValues(
+            gossip,
+            ring_offset,
+            chunks,
+        );
+    }
+
+    fn hasDuplicateForSlot(
+        gossip: *sig.gossip.GossipService,
+        my_pubkey: sig.core.Pubkey,
+        slot: Slot,
+    ) bool {
+        var gossip_table, var lock = gossip.gossip_table_rw.readWithLock();
+        defer lock.unlock();
```

Comment:
Why not just accept `gossip_table: *RwMux(sig.gossip.GossipTable)` here instead of passing in the entire gossip service, which otherwise goes unused besides the table?